Update RISC-V docs (#6376)
google/iree: docs/website/docs/building-from-source/riscv.md

# RISC-V cross-compilation

Running on a platform like RISC-V involves cross-compiling from a _host_
platform (e.g. Linux) to a _target_ platform (a specific RISC-V CPU architecture
and operating system):

* IREE's _compiler_ is built on the host and is used there to generate modules
  for the target
* IREE's _runtime_ is built on the host for the target. The runtime is then
  pushed to the target to run natively.

## Prerequisites

### Host environment setup

You should already be able to build IREE from source on your host platform.
Please make sure you have followed the [getting started](./getting-started.md)
steps.

### Install RISC-V cross-compile toolchain and emulator

You'll need a RISC-V LLVM compilation toolchain and a RISC-V enabled QEMU
emulator.

See instructions in the following links

* [Clang getting started](https://clang.llvm.org/get_started.html)
* [RISC-V GNU toolchain](https://github.com/riscv/riscv-gnu-toolchain)
* [QEMU](https://github.com/qemu/qemu)
* [RISC-V Linux QEMU](https://risc-v-getting-started-guide.readthedocs.io/en/latest/linux-qemu.html)

> !!! note
>     An environment variable `RISCV_TOOLCHAIN_ROOT` needs
>     to be set to the root directory of the installed GNU toolchain. The variable
>     can be used in building the RISCV target and a LLVM AOT module.

Reviewer: grammar nit: IMO the correct article would be "_an_ LLVM AOT module", if LLVM is pronounced "el el vee em" (which starts with a vowel sound). See also https://developers.google.com/style/articles.

To get around the grammar issue, you could reword so no article is needed:

```suggestion
!!! note
    The `RISCV_TOOLCHAIN_ROOT` environment variable needs to be set
    to the root directory of the installed GNU toolchain when building and
    using the RISCV compiler target.
```

(does it need to be set for both the build _and_ run steps, or just the run step for linker discovery?)

Author: It's definitely required for linker discovery, but it is also useful to set it for the target runtime compilation, as in the target configuration the CLI sets 
```shell
-DRISCV_TOOLCHAIN_ROOT=${RISCV_TOOLCHAIN_ROOT}
```

Author: Reword the note based on the comment.

### Install Prebuilt RISC-V Tools (RISC-V 64-bit Linux toolchain)

Execute the following script to download the prebuilt RISC-V toolchain and QEMU
from the IREE root directory:

```shell
./build_tools/riscv/riscv_bootstrap.sh
```
!!! note
    You also need to set `RISCV_TOOLCHAIN_ROOT`

>     (default at ${HOME}/riscv/toolchain/clang/linux/RISCV).

Reviewer: This is already mentioned just a few lines above. Omit here?

Author: Done.

```

### Target configuration

The following instruction shows how to build for a RISC-V 64-bit Linux machine.
For other RISC-V targets, please refer to
[riscv.toolchain.cmake](https://github.com/google/iree/blob/main/build_tools/cmake/riscv.toolchain.cmake)
as a reference of how to set up the cmake configuration.

#### RISC-V 64-bit Linux target
```shell
cmake -B ../iree-build-riscv/ \
  -DCMAKE_TOOLCHAIN_FILE="./build_tools/cmake/riscv.toolchain.cmake" \
  -DIREE_HOST_BINARY_ROOT=$(realpath ../iree-build-host/install) \
  -DRISCV_CPU=rv64 \
  -DIREE_BUILD_COMPILER=OFF \
  -DIREE_ENABLE_MLIR=OFF \
  -DRISCV_TOOLCHAIN_ROOT=${RISCV_TOOLCHAIN_ROOT}
  .
```

## Running IREE bytecode modules on the RISC-V system

!!! note
    The following instructions are meant for the RISC-V 64-bit Linux
    target. For the bare-metal target, please refer to
    [simple_embedding](https://github.com/google/iree/blob/main/iree/samples/simple_embedding)
    to see how to build a ML workload for a bare-metal machine.

Set the environment variable `RISCV_TOOLCHAIN_ROOT` if it is not set yet:

```shell
export RISCV_TOOLCHAIN_ROOT=<root directory of the RISC-V GNU toolchain>

> ```

Reviewer: Also mentioned yet again here :P

Author: Done.

Set the path to qemu-riscv64 Linux emulator binary in the `QEMU_BIN` environment
variable. If it is installed with `riscv_bootstrap.sh`, the path is default at
${HOME}/riscv/qemu/linux/RISCV/bin/qemu-riscv64.

```shell
export QEMU_BIN=<path to qemu-riscv64 binary>
```

> Invoke the host compiler tools produce input files:

Reviewer: Whoops, this doesn't make sense... and it is forked from https://google.github.io/iree/building-from-source/android/#running-tools-directly :/ 

Can you fix both places?

```suggestion
Invoke the host compiler tools to produce a FlatBuffer:
```

Author: Fixed this file and `android.md`

> # RISC-V cross-compilation

Author: Done.

> Running on a platform like RISC-V involves cross-compiling from a _host_

Reviewer: Are there some reference pages for RISC-V that would be good to include up here? I'm thinking of something like https://developer.android.com/ndk (a landing page that introduces the platform and links to further technical guides like https://developer.android.com/ndk/guides)

Looking around a bit, some ideas:

* https://wiki.riscv.org/display/TECH/RISC-V+Software+Ecosystem
* https://wiki.riscv.org/display/TECH
* https://riscv.org/

Author: external links on the toolchain and the emulator are added in the prerequisites section.

platform (e.g. Linux) to a _target_ platform (a specific RISC-V CPU architecture
and operating system):

* IREE's _compiler_ is built on the host and is used there to generate modules
  for the target
* IREE's _runtime_ is built on the host for the target. The runtime is then
  pushed to the target to run natively.

## Prerequisites

### Host environment setup

You should already be able to build IREE from source on your host platform.
Please make sure you have followed the [getting started](./getting-started.md)
steps.

### Install RISC-V cross-compile toolchain and emulator

You'll need a RISC-V LLVM compilation toolchain and a RISC-V enabled QEMU
emulator.

> * RISC-V toolchain is built from https://github.com/llvm/llvm-project (main branch).<br>

Reviewer: ... and now I'm realizing that should be updated to mention https://github.com/google/iree/blob/main/.github/workflows/publish_website.yml :P 

doc sources should be checked in to the repo on the `main` branch, and that action will automatically build the site and push it to https://github.com/google/iree/tree/gh-pages

to publish your changes on _your own GitHub pages_ site (e.g. to share a preview version during code review), you can use `mkdocs gh-deploy --remote-name <your remote>`

> * RISC-V toolchain is built from https://github.com/llvm/llvm-project (main branch).<br>

Author: I tried to deploy to my IREE fork, but stupidly used the same remo5e branch name `mkdocs gh-deploy --remote-name hcindyl --remote-branch rv-doc` and changed the PR review branch :( I reset the branch so the PR doesn't have the mkdocs commit.

> * RISC-V toolchain is built from https://github.com/llvm/llvm-project (main branch).<br>

Reviewer: :P don't blame yourself. It isn't clear enough that a "deploy" creates a separate directory tree

> * RISC-V toolchain is built from https://github.com/llvm/llvm-project (main branch).<br>

Author: Updated the section to put generic toolchain external links here.

  * Currently, the LLVM compiler is built on GNU toolchain, including libgcc,
    GNU linker, and C libraries. You need to build GNU toolchain first.<br>
  * Clone GNU toolchain from: https://github.com/riscv/riscv-gnu-toolchain

>     (master branch). Switch the "riscv-binutils" submodule to `rvv-1.0.x-zfh`

Reviewer: This instruction is very specific. Can we simplify this or link to an external guide instead?

Author: This is tricky as the vector extension instructions are not enabled in the master branch so to use the vector extension ops a specific `riscv-binutils` branch is used.

There's no public doc of the RVV support we can refer to yet. I can either add more explanation w.r.t the vector extension or not supporting the vector extension in this doc.

Author: Per offline discussion. Only provide the external links for the generic RISC-V toolchain support. Leave the `v` CPU feature out of this doc.

Reviewer: Another idea is to add a section at the bottom of the document for configuration options, possibly with a link from higher up. The main path through the document should be simple to follow, then dedicated users can go through the additional sections.

rough outline:
```
### Install RISC-V cross-compile toolchain and emulator

do X, Y, and Z
Note: for vector extensions support, see [link to header down in the page].

...

## Optional configuration

Vector extensions let SIMD code run more efficiently ... 

check out the `foo` branch on this repo and ...
```

Author: Optional section added at the end of the doc.

    branch manually.
* RISC-V QEMU is built from https://github.com/sifive/qemu/tree/v5.2.0-rvv-rvb-zfh

An environment variable `RISCV_TOOLCHAIN_ROOT` needs
to be set to the root directory of the installed GNU toolchain. The variable can
be used in building the RISCV target and a LLVM AOT module.

### Install Prebuilt RISC-V Tools (RISC-V 64-bit Linux toolchain)

Execute the following script to download the prebuilt RISC-V toolchain and QEMU:

```shell
# In IREE source root
$ ./build_tools/riscv/riscv_bootstrap.sh

> ```

Reviewer: Can you move the comment out of this code block and drop the leading `$`? (same for other code blocks)

---

I had been recommending this style (`# comments` and `$ echo "sample command"`) on the old documentation, but reversed my position on that with the new website.

The main reason is that mkdocs offers a "Copy to clipboard" button for code snippets that does not strip comments or leading prompt characters.

![image](https://user-images.githubusercontent.com/4010439/124155166-0b2f8c80-da4b-11eb-8032-eba5492b4194.png)

See also https://developers.google.com/style/code-syntax#prompt that _does_ recommend the leading `$`, but that website uses a widget that separates the code from the prompt: 
![image](https://user-images.githubusercontent.com/4010439/124156201-2fd83400-da4c-11eb-8112-6f8664932419.png)

What I really want is g3doc's feature set, but this isn't so bad.

Author: Done.

> **NOTE**:

Author: Done. Thanks for the suggestion.

* You also need to set `RISCV_TOOLCHAIN_ROOT`
(default at ${HOME}/riscv/toolchain/clang/linux/RISCV).

## Configure and build

### Host configuration

Build and install on your host machine:

``` shell
cmake -B ../iree-build/ \
  -DCMAKE_INSTALL_PREFIX=../iree-build/install \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  .
cmake --build ../iree-build/ --target install
```

### Target configuration

> Currently IREE verifies (and tests in CI) two RISC-V targets:

Reviewer: I'd like to be a bit careful with this sort of wording.

* "Currently ... two ... targets" in docs can get out of date with code easily
* I don't think we need to prove (in user documentation) how we achieve some technical capability. I'd rather just list which configurations are supported directly, perhaps with a list of known supported/not-yet-supported configurations like https://google.github.io/iree/#support-matrix (using https://squidfunk.github.io/mkdocs-material/reference/lists/#using-tasklists)

Author: Remove the "currently" and only note the instruction is for the RV64 Linux target

* RISC-V 64-bit CPU with Linux OS

> * RISC-V 32-bit CPU with "bare-metal" config -- a limited subset of IREE runtime

Reviewer: Structurally, a newline _should_ precede a bulleted list in markdown. This varies between markdown renderers (GitHub does not require a newline, g3doc does, mkdocs does).

![image](https://user-images.githubusercontent.com/4010439/124156643-b9880180-da4c-11eb-8365-7436372a9d1d.png)

* https://squidfunk.github.io/mkdocs-material/reference/lists/#using-unordered-lists

Author: Done. Thanks for catching this.

libraries that _can_ run on the bare-metal machine mode (with the machine BSP
linker script), but can also run on larger machines without the system library
support.

For other RISC-V targets, please refer to
[riscv.toolchain.cmake](https://github.com/google/iree/blob/main/build_tools/cmake/riscv.toolchain.cmake)
as a reference of how to set up the cmake configuration.

#### RISC-V 64-bit Linux target
```shell

> $ cmake -G Ninja -B ../iree-build-riscv/ \

Reviewer: nit: I've been omitting Ninja from sample commands on the website. There is a tip [on the getting started page](https://google.github.io/iree/building-from-source/getting-started/#clone-and-build) about Ninja, but we don't necessarily need to include it in every command.

Author: Ninja removed.

  -DCMAKE_TOOLCHAIN_FILE="./build_tools/cmake/riscv.toolchain.cmake" \
  -DIREE_HOST_BINARY_ROOT=$(realpath ../iree-build-host/install) \
  -DRISCV_CPU=rv64 \
  -DIREE_BUILD_COMPILER=OFF \
  -DIREE_ENABLE_MLIR=OFF \

>   -DIREE_BUILD_SAMPLES=ON \

Reviewer: nit: samples default to enabled and this document is not specific to a sample, so this option could be removed here

Author: Done.

  -DRISCV_TOOLCHAIN_ROOT=${RISCV_TOOLCHAIN_ROOT}
  .
```

For the RISC-V 32-bit bare-metal config, append the following CMake options
```shell
-DRISCV_CPU=rv32-baremetal \
-DIREE_BUILD_TESTS=OFF
```

## Running IREE bytecode modules on the RISC-V system

> **NOTE**:The following instructions are meant for the RISC-V 64-bit Linux

Reviewer: https://squidfunk.github.io/mkdocs-material/reference/admonitions/ instead of `**NOTE**`?

Author: Done. Thanks for the suggestion.

target. For the bare-metal target, please refer to
[simple_embedding](https://github.com/google/iree/blob/main/iree/samples/simple_embedding)
to see how to build a ML workload for a bare-metal machine.

Set the environment variable `RISCV_TOOLCHAIN_ROOT` if it is not set yet:

```shell
$ export RISCV_TOOLCHAIN_ROOT=<root directory of the RISC-V GNU toolchain>
```

Set the path to qemu-riscv64 Linux emulator binary in the `QEMU_BIN` environment
variable. If it is installed with `riscv_bootstrap.sh`, the path is default at
${HOME}/riscv/qemu/linux/RISCV/bin/qemu-riscv64.

```shell
$ export QEMU_BIN=<path to qemu-riscv64 binary>
```

Invoke the host compiler tools produce input files:

``` shell
$ ../iree-build/install/bin/iree-translate \
  -iree-mlir-to-vm-bytecode-module \
  -iree-hal-target-backends=vmvx \
  iree/samples/models/simple_abs.mlir \
  -o /tmp/simple_abs_vmvx.vmfb
```

Run the RISC-V emulation:

```shell
$ ${QEMU_BIN} \

>   -cpu rv64,x-v=true,x-k=true,vlen=256,elen=64,vext_spec=v1.0 \

Reviewer: Are these options worth explaining a bit? They look like magic to me :P

Maybe there's a reference page that could be linked? For Android, [this section](https://google.github.io/iree/building-from-source/android/#target-configuration) links to two external guides that explain the arguments (like this: https://developer.android.com/ndk/guides/cmake#variables)

Author: Same situation w.r.t supporting the vector extension instructions. This is the corresponding part of enabling the vector extension from the emulator side.

Author: Drop the vector extension support per offline conversion.

```shell
${QEMU_BIN} \
  -cpu rv64 \
  -L ${RISCV_TOOLCHAIN_ROOT}/sysroot/ \
  ../iree-build-riscv/iree/tools/iree-run-module \
  --driver=vmvx \
  --module_file=/tmp/simple_abs_vmvx.vmfb \
  --entry_function=abs \
  --function_input=f32=-5
```

## Optional configuration

[RISC-V Vector extensions](https://github.com/riscv/riscv-v-spec) allows SIMD
 code to run more efficiently. To enable the vector extension for the compiler
 toolchain and the emulator, build the tools from the following sources:

* RISC-V toolchain is built from
[https://github.com/llvm/llvm-project](https://github.com/llvm/llvm-project) (main branch).
  * Currently, the LLVM compiler is built on GNU toolchain, including libgcc,
    GNU linker, and C libraries. You need to build GNU toolchain first.
  * Clone GNU toolchain from:
  [https://github.com/riscv/riscv-gnu-toolchain](https://github.com/riscv/riscv-gnu-toolchain)
    (master branch). Switch the "riscv-binutils" submodule to `rvv-1.0.x-zfh`
    branch manually.
* RISC-V QEMU is built from
[https://github.com/sifive/qemu/tree/v5.2.0-rvv-rvb-zfh](https://github.com/sifive/qemu/tree/v5.2.0-rvv-rvb-zfh).

The SIMD code can be generated following the

> [IREE dynamic library CPU HAL driver flow](https://google.github.io/iree/deployment-configurations/cpu-dylib)

Reviewer: this can be a relative link

```suggestion
[IREE dynamic library CPU HAL driver flow](../deployment-configurations/cpu-dylib.md)
```

Author: Done.
